### PR TITLE
Make Solvers be FrozenObjects

### LINKED ISSUE #1044: Cache does not consider solver regularization
```
import nengo

nengo.cache.get_default_decoder_cache().invalidate()

with nengo.Network() as model_foo:
    x = nengo.networks.EnsembleArray(100, 1, seed=0)
    x.add_output('foo', function=None, solver=nengo.solvers.LstsqL2(reg=0.001))
    p_foo = nengo.Probe(x.foo, synapse=0.01)
    
with nengo.Simulator(model_foo) as sim_foo:
    sim_foo.run(1.0)

# nengo.cache.get_default_decoder_cache().invalidate()
    
with nengo.Network() as model_bar:
    x = nengo.networks.EnsembleArray(100, 1, seed=0)
    x.add_output('bar', function=None, solver=nengo.solvers.LstsqL2(reg=1000))
    p_bar = nengo.Probe(x.bar, synapse=0.01)
    
with nengo.Simulator(model_bar) as sim_bar:
    sim_bar.run(1.0)

print nengo.utils.numpy.rms(sim_foo.data[p_foo])
print nengo.utils.numpy.rms(sim_bar.data[p_bar])
```

Running this gives the output:
```
0.0329802255974
0.0329802255974
```

But if I uncomment the line in between to invalidate the cache:
```
0.0329802255974
3.57156978133e-07
```

The latter result is the correct one (since `bar` is supposed to be heavily regularized). This means that the cache is not key'ing on the solver's regularization.

### PR DESCRIPTION
**Pull request by hunse**
So they can take advantage of the built-in pickling and hashing.

This fixes #1044.

### CODE REVIEW

**Comment by Seanny123** [comment]
FYI, this passed all tests except for PEP8.

File: nengo/params.py [hunk 1]
```diff
@@ -416,7 +417,7 @@ def validate(self, instance, function_info):
         super(FunctionParam, self).validate(instance, function)
 
 
-class FrozenObject(object):
+class FrozenObject(with_metaclass(DocstringInheritor)):
```
Comment (tbekolay) [inline nengo/params.py]: Hm... are we sure we actually want this on the `FrozenObject` base class? I think it could be confusing if a user subclasses dist, for example, and doesn't write a docstring... calling `help` will give something wrong instead of blank.

The alternative is to do multiple inheritance in the solver; i.e., `class Solver(FrozenObject, with_metaclass(DocstringInheritor)):`.

File: nengo/utils/least_squares_solvers.py [hunk 2]
```diff
@@ -121,7 +131,11 @@ def __call__(self, A, Y, sigma, rng=None):
 class Conjgrad(LeastSquaresSolver):
     """Solve a least-squares system using conjugate gradient."""
 
+    tol = NumberParam('tol', low=0)
+    maxiters = IntParam('maxiters', low=1, optional=True)
+
```
Comment (tbekolay) [inline nengo/utils/least_squares_solvers.py]: Why is `X0` not included here?

File: nengo/utils/least_squares_solvers.py [hunk 3]
```diff
@@ -181,7 +195,10 @@ def _conjgrad_iters(calcAx, b, x, maxiters=None, rtol=1e-6):
 class BlockConjgrad(LeastSquaresSolver):
     """Solve a multiple-RHS least-squares system using block conj. gradient."""
 
+    tol = NumberParam('tol', low=0)
+
```
Comment (tbekolay) [inline nengo/utils/least_squares_solvers.py]: Ditto here with `X0`.

File: nengo/utils/least_squares_solvers.py [hunk 4]
```diff
@@ -253,7 +270,11 @@ class RandomizedSVD(LeastSquaresSolver):
     ``sklearn.utils.extmath.randomized_svd`` for details about the parameters.
     """
 
+    n_components = IntParam('n_components', low=1)
+    n_oversamples = IntParam('n_oversamples', low=0)
+
```
Comment (tbekolay) [inline nengo/utils/least_squares_solvers.py]: Should add a `DictParam` for the kwargs too? In which case, it should maybe be passed in as an explicit dict, rather than as `**kwargs`?

**Comment by tbekolay** [comment]
Made some inline comments; other than those this is a definite improvement!

File: nengo/params.py (hunk 1 above)
Comment (hunse) [inline nengo/params.py]: Multiple inheritance was my original plan, but I was having some trouble getting it working (can't remember what). If you can do it, I'm all for it.

Comment (tbekolay) [inline nengo/params.py]: Will give it a shot!

File: nengo/utils/least_squares_solvers.py (hunk 2 above)
Comment (hunse) [inline nengo/utils/least_squares_solvers.py]: I had a reason... maybe because it could be big and I didn't want to worry about hashing/pickling it? Also, it's kind of a strange parameter in general because it has to be the right size for the system being solved, meaning that when it's set the solver will only work for one size of connection. But maybe it should be included...

File: nengo/utils/least_squares_solvers.py (hunk 4 above)
Comment (hunse) [inline nengo/utils/least_squares_solvers.py]: Does `DictParam` have a proper hashing function right now? Maybe we should just add all the possible parameters manually. There aren't too many of them.

File: nengo/params.py (hunk 1 above)
Comment (jgosmann) [inline nengo/params.py]: > The alternative is to do multiple inheritance in the solver; i.e., class Solver(FrozenObject, with_metaclass(DocstringInheritor)):.

Shouldn't it be `Solver(with_metaclass(DocstringInheritor, FrozenObject)):`? At least I believe that is what I used in some of my PRs.

Comment (hunse) [inline nengo/params.py]: From what I vaguely know about metaclasses, I think that would make sense. Maybe that's why it wasn't working for me before. I definitely had it like @tbekolay did above.

Comment (tbekolay) [inline nengo/params.py]: All 3 possibilities (`with_metaclass(DocstringInheritor), FrozenObject`; `FrozenObject, with_metaclass(DocstringInheritor)`; `with_metaclass(DocstringInheritor, FrozenObject)`) passed unit tests for me, but @jgosmann's is undoubtedly the better solution (results in a shorter `mro`, I think), so I pushed a fixup commit with that.

Comment (hunse) [inline nengo/params.py]: Did you check that the docstrings are being properly inherited, though? There's no test for that, and I think that's what wasn't working for me.

Comment (hunse) [inline nengo/params.py]: I just checked, and it seems to be working this way, so it's all good!

Comment (hunse) [inline nengo/params.py]: It also seems to work the other ways, so I'm not sure what I was doing wrong before...

Comment (tbekolay) [inline nengo/params.py]: Sometimes we just have to have faith in the gods of programming. Surely they have a glorious plan for us!

File: nengo/solvers.py [hunk 5]
```diff
@@ -67,16 +72,14 @@ def mul_encoders(self, Y, E, copy=False):
...
-                    "Encoders must be 'None' for decoder solver", attr='E')
-            return Y.copy() if copy else Y
+        if self.weights and E is None:
+            raise ValidationError(
+                "Encoders must be provided for weight solver", attr='E')
+        if not self.weights and E is not None:
+            raise ValidationError(
+                "Encoders must be 'None' for decoder solver", attr='E')
+
+        return np.dot(Y, E) if E is not None else Y.copy() if copy else Y
```
Comment (tbekolay) [inline nengo/solvers.py]: The nested conditionals here makes it a bit hard to parse... could you put `Y.copy() if copy else Y` in parentheses?

Comment (hunse) [inline nengo/solvers.py]: That's interesting, I don't really consider this nested, and feel like it reads exactly like the logic goes. That's one thing I like about Python: the statement `x if a else y if b else z if c else q` does exactly what you think it would do if someone said it to you (at least to me). If those parentheses would be helpful, though, I'm fine to add them.

Comment (tbekolay) [inline nengo/solvers.py]: Nah, not a critical thing, I see your point! I think for me, I parse these lines with an eye to the value, rather than the condition... so when I see `x if a else y if b else z if c else q` I'm thinking about what is the situation in which you get `b`, rather than reading it left to right. So I'm looking at `q` for example, then I scan left and see that `q` happens if `a` `b` and `c` are False, which takes a few seconds to think through. For some reason the parens help with the mental organization of that, but like I said, not a critical thing so no need to change.

**Comment by jgosmann** [comment]
This makes the tests fail for some reason?

**Comment by tbekolay** [comment]
Hmm... concerning that that didn't fail [pylint](https://travis-ci.org/nengo/nengo/jobs/130609753)

**Comment by hunse** [comment]
Ok, hopefully those two commits fix everything.

**Comment by jgosmann** [comment]
> Hmm... concerning that that didn't fail pylint

I assume pylint isn't able to tell the the super call is needed, because of the `with_metaclass` inheritance.

**Comment by hunse** [comment]
I think this is ready to go.

**Comment by tbekolay** [comment]
OK, I'll look at this now

File: nengo/utils/least_squares_solvers.py [hunk 6]
```diff
@@ -142,14 +143,15 @@ def __init__(self, tol=1e-2, maxiters=None, X0=None):
 
     def __call__(self, A, Y, sigma, rng=None):
         Y, m, n, d, matrix_in = format_system(A, Y)
+        X = np.zeros((n, d)) if self.X0 is None else np.array(self.X0)
+        if X.shape != (n, d):
+            raise ValueError('X0 must be shape %s, got %s' % ((n, d), X.shape))
```
Comment (tbekolay) [inline nengo/utils/least_squares_solvers.py]: Should be a ValidationError rather than ValueError

File: nengo/utils/least_squares_solvers.py [hunk 7]
```diff
@@ -207,14 +210,16 @@ def __call__(self, A, Y, sigma, rng=None):
         sigma = np.asarray(sigma, dtype='float')
         sigma = sigma.reshape(sigma.size, 1)
 
+        X = np.zeros((n, d)) if self.X0 is None else np.array(self.X0)
+        if X.shape != (n, d):
+            raise ValueError('X0 must be shape %s, got %s' % ((n, d), X.shape))
```
Comment (tbekolay) [inline nengo/utils/least_squares_solvers.py]: ValidationError here too

**Comment by tbekolay** [comment]
I'll do the `ValueError` -> `ValidationError` change in the merge!
